### PR TITLE
feat: add OAuth support

### PR DESCRIPTION
**Pull request by PeterSchafer**
### Description

_Provide description of this PR and changes, if linked Jira ticket doesn't cover it in full._

### Checklist

- [ ] Tests added and all succeed
- [ ] Linted
- [ ] CHANGELOG.md updated
- [ ] README.md updated, if user-facing

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs [hunk 1]
```diff
@@ -38,21 +41,32 @@ public class SnykGeneralOptionsDialogPage : DialogPage, ISnykOptions
...
+                return;
+            }
+
+            var apiEndpointResolver = new ApiEndpointResolver(this);
+            var type = apiEndpointResolver.AuthenticationMethod;
+            this.apiToken = new AuthenticationToken(type, apiToken);
+
+            this.FireSettingsChangedEvent();
+        }
+
```
[the remark this answers is not in the document]
Comment (asaf92) [reply Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs]: (never mind, the token type is different)

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs [hunk 2]
```diff
@@ -38,19 +48,41 @@ public class SnykGeneralOptionsDialogPage : DialogPage, ISnykOptions
...
-                this.apiToken = value;
-                this.FireSettingsChangedEvent();
+        public void SetApiToken(string apiToken)
+        {
+            if (this.apiToken?.ToString() == apiToken)
+            {
+                return;
             }
+
+            this.apiToken = CreateAuthenticationToken(apiToken);
```
Comment (PeterSchafer) [inline Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs:68]: nitpick: possible to call SetApiToken(AuthenticationToken token) instead duplicating code

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs [hunk 3]
```diff
@@ -244,7 +276,64 @@ public void Initialize(ISnykServiceProvider provider)
...
+            Logger.Information("Enter Authenticate method");
+            var cli = this.ServiceProvider.NewCli();
+            if (!cli.IsCliFileFound())
+            {
+                ThrowFileNotFoundException();
+            }
+            try
+            {
+                // Pull token from configuration. If the token is invalid, attempt to authenticate and get a new token.
+                var apiTokenString = ServiceProvider.NewCli().GetApiToken();
```
Comment (PeterSchafer) [inline Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs:292]: question: any reason why this doesn't use the cli instance from above?

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs (hunk 2 above)
Comment (asaf92) [reply Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs:68]: Done

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs (hunk 3 above)
Comment (asaf92) [reply Snyk.VisualStudio.Extension.Shared/Settings/SnykGeneralOptionsDialogPage.cs:295]: Nope. I don't even see any reason why the CLI isn't a singleton.

Changed it to reuse the CLI instance from above

**Review by asaf92 — APPROVED** [review]

**Pull request merged**